_adr/92/index.adoc

```diff
@@ -0,0 +1,108 @@
+
+image::with_BFF.png[Proposed Architecture]
+
+The service is implemented using Java and Quarkus to achieve maximum collaboration across the current teams and align with Red Hat tech strategy.
+The API layer is designed using API Designer(RHOAD) and conforming to the OpenAPI 3+ specification to achieve maximum collaboration across different teams and members with various technological background.
+The development is driven by a "Contract First" approach leveraging the [Apicurio Codegen Quarkus Extension](https://github.com/Apicurio/apicurio-codegen/tree/main/quarkus-extension) and re-using the Java SDKs(based on Kiota) provided by the DevExp Team.
+The BFF have to to be stateless so that it can easily be scaled up and down according to the needs.
+All of the exposed endpoints are going to perform 1+ REST APIs calls to the underlying services and provide back the result.
+Access to the underlying services will be done by propagating the original(user) access token to simplify development and to keep the security model similar to the current architecture.
+
+### Threat model
+
+Being it a publicly accessible service the generic service Threat model applies, we should ensure stability of the platform by having all the good practices in place:
+
+* Rate Limiting
+* Circuit Breaking
+* Caching
+
+Regarding multitenancy, it can be achieved, in the first prototype by re-using the Access Token for the initial API request to issue the following requests to the backend services.
+
+## Alternatives Considered / Rejected
+
+### GraphQL
+
+Is a well known technical platform for implementing BFFs, especially strong for composing data from multiple services but it doesn't magically solves the problem by itself.
+When comparing a possible OpenAPI solution to a GraphQL one, for this specific problem, the real advantage in favor of the latter is the possibility for dynamic fields selections in the client.
+We do not consider this upside strong enough to outweighs some of the challenges:
+
+* Lack of expertise in the technology - one of the objective of this component is to enable multiple teams collaboration on the repo
+* Challenges with Observability - Support for partial failures compromise the usage of "generic" REST APIs monitoring tools
+
+We leave the doors open to re-evaluate the choice of GraphQL in the future and/or in different contexts.
+
+### Expand the current services APIs
+
+We want the current microservices to stay vertical and opinionated toward a minimal and curated API that directly maps the functionality of the service without having to deal with the integration challenges.
+
+## Challenges
+
+* Caching while guaranteeing multitenancy
+* Shared ownership of the component across different teams
+* Avoid API proliferation after the first BFF is in place
+* Potential lack of control over APIs being aggregated
+* The introduction of private data planes can introduce additional issues because of backend APIs reachability
```

Comment:
This would seem to be a real problem. Not supporting private clusters seems like quite the architectural draw back given we want to address the enterprise use case?

Reply:
I think the bulk of the issue relies on the control plane, which requires clients to connect and manipulate many responses from different services. We bleed a lot of architectural decisions and problems to clients.

The data plane is easier; clients connect with the instance and work with the instance directly. There is less need to connect with other services (as of today), so I think we can keep clients connecting with the data plane directly.

Still, the lessons learned can and should be applied to the data plane for future developments.